### PR TITLE
Feature/d3 asim 1320

### PR DESCRIPTION
**Pull request by spyrostz**

### CODE REVIEW

File: src/d3a/models/strategy/__init__.py [hunk 1]
```diff
@@ -56,78 +56,89 @@ class Offers:
...
+        self.posted = {}  # type: Dict[Offer, Str]
+        self.sold = {}  # type: Dict[Str, List[str]]
         self.changed = {}  # type: Dict[str, Offer]
 
+    @property
+    def area(self):
+        # TODO: Remove the owner and area distinction from the AreaBehaviorBase class
+        return self.strategy.area if self.strategy.area is not None else self.strategy.owner
+
+    def filter_offers_in_future_markets(self, existing_offers):
```
Comment (hannesdiedrich) [inline src/d3a/models/strategy/__init__.py]: AFAIU, this function deletes all "past" offers right? Very minor, but if so, can you please rename the function accordingly, maybe to "delete_past_offers" or so? Thanks!

File: src/d3a/models/strategy/__init__.py [hunk 2]
```diff
@@ -247,65 +260,70 @@ def post_bid(self, market, price, energy):
...
             "Invalid state, cannot receive a bid if single sided market is globally configured."
-        market = self.area.get_future_market_from_id(market_id)
-        assert market is not None
 
         if bid_trade.buyer == self.owner.name:
-            self.add_bid_to_bought(bid_trade.offer, self.area.get_future_market_from_id(market_id))
+            self.add_bid_to_bought(bid_trade.offer, market_id)
+
+    def event_market_cycle(self):
+        from d3a.models.const import ConstSettings
```
Comment (hannesdiedrich) [inline src/d3a/models/strategy/__init__.py]: Is this due to circular import or can this be moved to the top?

File: src/d3a/models/strategy/area_agents/one_sided_alternative_pricing_agent.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ def _buy_energy_alternative_pricing_schemes(self, area):
                     else:
                         raise MarketException
                     offer.price = offer.energy * sell_rate
-                    self.accept_offer(offer.market, offer)
+                    self.accept_offer(self.lower_market, offer)
```
Comment (hannesdiedrich) [inline src/d3a/models/strategy/area_agents/one_sided_alternative_pricing_agent.py]: Thanks for fixing!

File: src/d3a/models/area/event_dispatcher.py [hunk 4]
```diff
@@ -75,9 +75,12 @@ def _broadcast_notification(self, event_type: Union[MarketEvent, AreaEvent], **k
 
             if not self.area.events.is_connected:
                 break
+            # TODO: Replace agents with just one agent, since one object is
+            # there fore every market slot
```
Comment (hannesdiedrich) [inline src/d3a/models/area/event_dispatcher.py]: Yes! Shall we create a story for this?

File: src/d3a/models/area/event_dispatcher.py [hunk 5]
```diff
@@ -175,3 +181,19 @@ def create_area_agents(self, is_spot_market, market):
         if self.area.parent:
             # Add inter area appliance to report energy
             self.area.appliance = InterAreaAppliance(self.area.parent, self.area)
+
+    def _delete_past_agents(self, timeslot, area_agent_member):
+        if not ConstSettings.GeneralSettings.KEEP_PAST_MARKETS:
+            delete_agents = [pm for pm in area_agent_member.keys() if
+                             self.area.current_market and pm < self.area.current_market.time_slot]
+            for pm in delete_agents:
+                for i in area_agent_member[pm]:
```
Comment (hannesdiedrich) [inline src/d3a/models/area/event_dispatcher.py]: Can you please exchange `i` with `area_agent` and `j` with engine or something similar and more verbose. Thanks!

File: src/d3a/models/area/event_dispatcher.py (hunk 4 above)
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py]: Thanks for pointing out, I should delete this comment. There are unfortunately more than one agents even now, in order to accommodate the future markets, therefore we cannot do that unfortunately.

File: src/d3a/models/area/event_dispatcher.py (hunk 5 above)
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py]: will do, thanks!

File: src/d3a/models/strategy/__init__.py (hunk 2 above)
Comment (spyrostz) [reply src/d3a/models/strategy/__init__.py]: Let me try, I think it is possible to move it to the top, thanks!

File: src/d3a/models/strategy/__init__.py (hunk 1 above)
Comment (spyrostz) [reply src/d3a/models/strategy/__init__.py]: delete_past_offers sounds great. I will also rename accordingly the delete_past_markets method to delete_past_markets_offers .

**Pull request merged**